sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                                          str(gap) + 'sec\nend_prev: ' + _from_last + '"]' + '[ fontcolor="' + color + '" color=' + color + ']'
-                                          ))
-
-            for vname, signatures in nodes.items(): # Go over all vertices again and define their shapes + make high-sev sink states dotted
-                mas = vname.split('|')[0]
-                mas = macro_inv[micro2macro['MicroAttackStage.'+mas]]
-                shape = shapes[mas]
-                if shape == shapes[0] or vname.split('|')[2] in already_addressed: # if it's oval, we dont do anything because its not high-sev sink
-                    lines.append((0,'"'+translate(vname)+'" [shape='+shape+']'))
+                                partial = '"' + _translate(vname) + '" [style="dotted'  # Redefine here
+                                if not sum([True if partial in line else False for line in lines]):
+                                    lines.append(partial + '"]')
+
+                    # Create edges (transitions) for the AG
+                    bigram = zip(attempt, attempt[1:])  # Make bigrams (sliding window of 2)
+                    for vid, ((vname1, _, _, _, ts1), (vname2, _, _, _, ts2)) in enumerate(bigram):
+                        edge_line = _create_edge_line(vid, vname1, vname2, ts1, ts2, color, team_attacker.split('-')[1])
+                        lines.append(edge_line)
+
+            # Go over all vertices again and define their shapes + make high-sev sink states dotted
+            for vname, signatures in node_signatures.items():
+                mcat = vname.split('|')[0]
+                mcat = macro_inv[micro2macro['MicroAttackStage.' + mcat]]
+                shape = shapes[mcat]
+                # If it's oval, we don't do anything because it is not a high-sev sink
+                if shape == shapes[0] or vname.split('|')[2] in already_addressed:
+                    lines.append('"' + _translate(vname) + '" [shape=' + shape + ']')
                 else:
-                    sinkflag = False
-                    for sink in sev_sinks:    
-                        if vname.split("|")[-1] == sink:
-                            sinkflag = True
-                            break
-                    if sinkflag:
-                        lines.append((0,'"'+translate(vname)+'" [style="dotted", shape='+shape+']'))
+                    if _is_severe_sink(vname, sev_sinks):
+                        lines.append('"' + _translate(vname) + '" [style="dotted", shape=' + shape + ']')
                     else:
-                        lines.append((0,'"'+translate(vname)+'" [shape='+shape+']'))
-                # add tooltip
-                lines.append((1, '"'+translate(vname)+'"'+' [tooltip="'+ "\n".join(signatures) +'"]'))
-            lines.append((1000,'}'))
-            
-            for l in lines: # count vertices and edges
```

Comment:
I think this part was used in the main paper, so I have kept it, but I have extracted it in a separate method, so that it is more readable and only one line has to be commented out